cartoframes/dataset.py

```diff
@@ -24,51 +33,74 @@ class Dataset(object):
     PUBLIC = 'public'
     LINK = 'link'
 
+    TABLE_TYPE = 'Table'
+    QUERY_TYPE = 'Query'
+    DATAFRAME_TYPE = 'DataFrame'
+    GEODATAFRAME_TYPE = 'GeoDataFrame'
+
     DEFAULT_RETRY_TIMES = 3
 
-    def __init__(self, carto_context, table_name, schema='public', df=None):
-        self.cc = carto_context
-        self.table_name = normalize_name(table_name)
+    def __init__(self, type, data, context=None, schema='public', table_name=None):
+        self.type = type
+        self.data = data
         self.schema = schema
-        self.df = df
-        if self.df is not None:
-            self.normalized_column_names = _normalize_column_names(self.df)
-        warn('Table will be named `{}`'.format(table_name))
-
-    @staticmethod
-    def create_from_query(cart_context, query, table_name):
-        dataset = Dataset(cart_context, table_name)
-        dataset.cc.batch_sql_client \
-               .create_and_wait_for_completion(
-                   '''BEGIN; {drop}; {create}; {cartodbfy}; COMMIT;'''
-                   .format(drop=dataset._drop_table_query(),
-                           create=dataset._create_table_from_query(query),
-                           cartodbfy=dataset._cartodbfy_query()))
-
-        return dataset
+        self.cc = context or default_context
 
-    def upload(self, with_lonlat=None, if_exists='fail'):
-        if self.df is None:
-            raise ValueError('You have to create a `Dataset` with a pandas DataFrame in order to upload it to CARTO')
+        if type == Dataset.TABLE_TYPE:
+            self.table_name = normalize_name(data)
+            if self.table_name != data:
+                warn('Table will be named `{}`'.format(data))
+        elif table_name:
+            self.table_name = normalize_name(table_name)
+            if self.table_name != table_name:
+                warn('Table will be named `{}`'.format(table_name))
 
-        if not self.exists():
-            self._create_table(with_lonlat)
-        else:
-            if if_exists == Dataset.FAIL:
-                raise NameError(('Table with name {table_name} already exists in CARTO.'
-                                 ' Please choose a different `table_name` or use'
-                                 ' if_exists="replace" to overwrite it').format(table_name=self.table_name))
-            elif if_exists == Dataset.REPLACE:
-                self._create_table(with_lonlat)
+        if type == Dataset.DATAFRAME_TYPE or type == Dataset.GEODATAFRAME_TYPE:
```

Comment:
Maybe we can have a constant at the class level 

```python
DATAFRAME_TYPES = [Dataset.DATAFRAME_TYPE, Dataset.GEODATAFRAME_TYPE]
```

and refactor this (and other conditions like this) as:

```python
if self.type in DATAFRAME_TYPES
```